# SQL Final Project

### Project Description

The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were at home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers.

You've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product. 

### Goals of the study (required queries)

* Find the number of books released after January 1, 2000.
* Find the number of user reviews and the average rating for each book.
* Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
* Identify the author with the highest average book rating (look only at books with at least 50 ratings).
* Find the average number of text reviews among users who rated more than 50 books.

# Tables Of Contents <a class="anchor" id="table_of_contents"></a>

* [Import Libraries](#chapter1)
* [Access to the DB](#chapter2)
* [Table Queries](#chapter3)
* [Taking a look at the data](#chapter4)
    * [Books](#chapter4.1)
    * [Authors](#chapter4.2)
    * [Publishers](#chapter4.3)
    * [Ratings](#chapter4.4)
    * [Reviews](#chapter4.5)
* [SQL Queries](#chapter5)
    * [Q1](#chapter5.1)
    * [Q2](#chapter5.2)
    * [Q3](#chapter5.3)
    * [Q4](#chapter5.4)
    * [Q5](#chapter5.5)
* [General Conclusion](#chapter6)


### Import Libraries<a class="anchor" id="chapter1"></a>

In [1]:
import pandas as pd
from sqlalchemy import create_engine
import psycopg2
import warnings
warnings.filterwarnings('ignore')

### Access to the DB<a class="anchor" id="chapter2"></a>

In [2]:
db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Table Queries<a class="anchor" id="chapter3"></a>

Function that takes a query and return dataframe for general use.

In [3]:
def queryResult(q):
    return pd.io.sql.read_sql(q, con = engine)

### Taking a look at the data <a class="anchor" id="chapter4"></a>

In [4]:
engine.table_names()

['orders',
 'visits',
 'advertisment_costs',
 'trip',
 'C:\x08oks',
 'boks',
 'finished_lessons1',
 'finished_lessons2',
 'publishers',
 'authors',
 'reviews',
 'ratings',
 'finished_lessons',
 'books']

**Books**<a class="anchor" id="chapter4.1"></a>

In [5]:
query = '''
           SELECT 
            *
           FROM
            books
        '''
queryResult(query).head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

**Authors**<a class="anchor" id="chapter4.2"></a>

In [6]:
query = '''
           SELECT 
            *
           FROM
            authors
        '''
queryResult(query).head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

**Publishers**<a class="anchor" id="chapter4.3"></a>

In [7]:
query = '''
           SELECT 
            *
           FROM
            publishers
        '''
queryResult(query).head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

**Ratings**<a class="anchor" id="chapter4.4"></a>

In [8]:
query = '''
           SELECT 
            *
           FROM
            ratings
        '''
queryResult(query).head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

**Reviews**<a class="anchor" id="chapter4.5"></a>

In [9]:
query = '''
           SELECT 
            *
           FROM
            reviews
        '''
queryResult(query).head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

# SQL Queries<a class="anchor" id="chapter5"></a>

**Find the number of books released after January 1, 2000.**<a class="anchor" id="chapter5.1"></a>

In [10]:
query = '''
    SELECT
        COUNT(DISTINCT book_id) AS books_amount
    FROM
        books
    WHERE
        books.publication_date > '2000-01-01'
        '''
queryResult(query)

books_amount
0           819

819 books from the database released after January 1, 2000.

**Find the number of user reviews and the average rating for each book.**<a class="anchor" id="chapter5.2"></a>

In [11]:
query = '''
    SELECT
        books.title AS title,
        books.book_id,
        COUNT(DISTINCT reviews.review_id) AS review_amount,
        AVG(ratings.rating) AS avg_rating
    FROM
        books
    LEFT JOIN reviews ON 
        books.book_id = reviews.book_id
    LEFT JOIN ratings ON 
        books.book_id = ratings.book_id
    GROUP BY
        books.book_id
    ORDER BY
        review_amount DESC
        '''
queryResult(query)

title  book_id  \
0                              Twilight (Twilight  #1)      948   
1                                  Water for Elephants      963   
2                                     The Glass Castle      734   
3    Harry Potter and the Prisoner of Azkaban (Harr...      302   
4    The Curious Incident of the Dog in the Night-Time      695   
..                                                 ...      ...   
995    Anne Rice's The Vampire Lestat: A Graphic Novel       83   
996                            The Natural Way to Draw      808   
997   The Cat in the Hat and Other Dr. Seuss Favorites      672   
998                          Essential Tales and Poems      221   
999  Disney's Beauty and the Beast (A Little Golden...      191   

     review_amount  avg_rating  
0                7    3.662500  
1                6    3.977273  
2                6    4.206897  
3                6    4.414634  
4                6    4.081081  
..             ...         ...  
995              0    3.666667  
996              0    3.000000  
997              0    5.000000  
998              0    4.000000  
999              0    4.000000  

[1000 rows x 4 columns]

We can see the average rating and amount of reviews for each book.

**Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).**<a class="anchor" id="chapter5.3"></a>

In [12]:
query = '''
    SELECT
        publishers.publisher_id,
        publishers.publisher AS name,
        COUNT(DISTINCT book_id) AS books_amount
    FROM
        publishers
    JOIN books ON 
        publishers.publisher_id = books.publisher_id
    WHERE
        books.num_pages > 50
    GROUP BY
        publishers.publisher_id, publishers.publisher 
    ORDER BY
        books_amount DESC
    LIMIT
        1
        '''
queryResult(query)

publisher_id           name  books_amount
0           212  Penguin Books            42

The publisher Penguin Books, with the publisher_id of 212, released the greatest number of books with more than 50 pages - 42 books.

**Identify the author with the highest average book rating (look only at books with at least 50 ratings).**<a class="anchor" id="chapter5.4"></a>

Version 1: Highest average rating per author for all his books with at least 50 ratings. 

In [13]:
query = '''
    SELECT
        author_id,
        author,
        AVG(avg_rating) AS avg_rating
    FROM
        (SELECT
            authors.author_id,
            authors.author,
            ratings.book_id,
            AVG(ratings.rating) AS avg_rating
        FROM
            books
        JOIN authors ON 
            books.author_id = authors.author_id
        JOIN ratings ON 
            books.book_id = ratings.book_id
        GROUP BY
            authors.author_id,
            authors.author,
            ratings.book_id
        HAVING
            COUNT(ratings.rating_id) > 49) AS subq
    GROUP BY
        author_id,
        author
    ORDER BY
        avg_rating DESC
    LIMIT
        1
        '''
queryResult(query)

author_id                      author  avg_rating
0        236  J.K. Rowling/Mary GrandPré    4.283844

Version 2: Highest average rating per book for books with at least 50 ratings. 

In [14]:
query = '''
    SELECT
        authors.author_id,
        authors.author,
        ratings.book_id,
        AVG(ratings.rating) AS avg_rating
    FROM
        books
    JOIN authors ON 
        books.author_id = authors.author_id
    JOIN ratings ON 
        books.book_id = ratings.book_id
    GROUP BY
        authors.author_id,
        authors.author,
        ratings.book_id
    HAVING
        COUNT(ratings.rating_id) > 49
    ORDER BY
        avg_rating DESC
    LIMIT
        1
        '''
queryResult(query)

author_id                      author  book_id  avg_rating
0        236  J.K. Rowling/Mary GrandPré      302    4.414634

The author J.K. Rowling/Mary GrandPré, with the author_id of 236 is the author with the highest average book rating.

**Find the average number of text reviews among users who rated more than 50 books.**<a class="anchor" id="chapter5.5"></a>

In [15]:
query = '''
    SELECT
        AVG(count) AS avg_review_amount
    FROM
        (SELECT
            username,
            COUNT (DISTINCT review_id)
        FROM
            reviews
        WHERE
            username IN (SELECT
                            username
                        FROM
                            ratings
                        GROUP BY
                            username
                        HAVING
                            COUNT(DISTINCT book_id) > 50)
        GROUP BY
            username) s
        '''
queryResult(query)

avg_review_amount
0          24.333333

Among users who rated more than 50 books, the average number of text reviews is 24.333.

# General Conclusion<a class="anchor" id="chapter6"></a>

We studied the tables, made an SQL query for each of the tasks and described our conclusions for each of the tasks. This information will help us generate a value proposition for a new product. 